[SPARK-23435][SPARKR][TESTS] Update testthat to >= 2.0.0

zero323 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->

- Update `testthat` to >= 2.0.0
- Replace of `testthat:::run_tests` with `testthat:::test_package_dir`
- Add trivial assertions for tests, without any expectations, to avoid skipping.
- Update related docs.


### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->

`testthat` version has been frozen by [SPARK-22817](https://issues.apache.org/jira/browse/SPARK-22817) / https://github.com/apache/spark/pull/20003, but 1.0.2 is pretty old, and we shouldn't keep things in this state forever.

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->

No.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

- Existing CI pipeline:
     - Windows build on AppVeyor, R 3.6.2, testthtat 2.3.1
     - Linux build on Jenkins, R 3.1.x, testthat 1.0.2 

- Additional builds with thesthat 2.3.1  using [sparkr-build-sandbox](https://github.com/zero323/sparkr-build-sandbox) on c7ed64af9e697b3619779857dd820832176b3be3 

   R 3.4.4  (image digest ec9032f8cf98)
   ```
   docker pull zero323/sparkr-build-sandbox:3.4.4
   docker run zero323/sparkr-build-sandbox:3.4.4 zero323 --branch SPARK-23435 --commit c7ed64af9e697b3619779857dd820832176b3be3 --public-key https://keybase.io/zero323/pgp_keys.asc
    ```
    3.5.3 (image digest 0b1759ee4d1d)

    ```
    docker pull zero323/sparkr-build-sandbox:3.5.3
    docker run zero323/sparkr-build-sandbox:3.5.3 zero323 --branch SPARK-23435 --commit 
    c7ed64af9e697b3619779857dd820832176b3be3 --public-key https://keybase.io/zero323/pgp_keys.asc
    ```

   and 3.6.2 (image digest 6594c8ceb72f)
    ```
   docker pull zero323/sparkr-build-sandbox:3.6.2
   docker run zero323/sparkr-build-sandbox:3.6.2 zero323 --branch SPARK-23435 --commit c7ed64af9e697b3619779857dd820832176b3be3 --public-key https://keybase.io/zero323/pgp_keys.asc
   ````


   Corresponding [asciicast](https://asciinema.org/) are available as 10.5281/zenodo.3629431

     [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3629431.svg)](https://doi.org/10.5281/zenodo.3629431) 

   (a bit to large to burden asciinema.org, but can run locally via `asciinema play`).


----------------------------

Continued from #27328

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22145/
Test PASSed.

SparkQA commented:
**[Test build #117385 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117385/testReport)** for PR 27359 at commit [`28115ce`](https://github.com/apache/spark/commit/28115ce6967d9e4d3db2cc4086d7f60d1874ba4c).

SparkQA commented:
**[Test build #117385 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117385/testReport)** for PR 27359 at commit [`28115ce`](https://github.com/apache/spark/commit/28115ce6967d9e4d3db2cc4086d7f60d1874ba4c).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117385/
Test FAILed.

SparkQA commented:
**[Test build #117389 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117389/testReport)** for PR 27359 at commit [`2e265ab`](https://github.com/apache/spark/commit/2e265abe386a24b9c824da53dfd607587202c46f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22149/
Test PASSed.

SparkQA commented:
**[Test build #117389 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117389/testReport)** for PR 27359 at commit [`2e265ab`](https://github.com/apache/spark/commit/2e265abe386a24b9c824da53dfd607587202c46f).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117389/
Test FAILed.

SparkQA commented:
**[Test build #117392 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117392/testReport)** for PR 27359 at commit [`5a565fa`](https://github.com/apache/spark/commit/5a565fa1ecd61096c20e16751cb5ea7af4815659).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22152/
Test PASSed.

SparkQA commented:
**[Test build #117392 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117392/testReport)** for PR 27359 at commit [`5a565fa`](https://github.com/apache/spark/commit/5a565fa1ecd61096c20e16751cb5ea7af4815659).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117392/
Test FAILed.

SparkQA commented:
**[Test build #117395 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117395/testReport)** for PR 27359 at commit [`c4cd053`](https://github.com/apache/spark/commit/c4cd05343523556676dfea1747a942b7bd733c98).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22155/
Test PASSed.

SparkQA commented:
**[Test build #117395 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117395/testReport)** for PR 27359 at commit [`c4cd053`](https://github.com/apache/spark/commit/c4cd05343523556676dfea1747a942b7bd733c98).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117395/
Test FAILed.

SparkQA commented:
**[Test build #117397 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117397/testReport)** for PR 27359 at commit [`cb5bf59`](https://github.com/apache/spark/commit/cb5bf5930ae22822357ea176497dac58d4bd1a9e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22157/
Test PASSed.

SparkQA commented:
**[Test build #117397 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117397/testReport)** for PR 27359 at commit [`cb5bf59`](https://github.com/apache/spark/commit/cb5bf5930ae22822357ea176497dac58d4bd1a9e).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117397/
Test FAILed.

SparkQA commented:
**[Test build #117399 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117399/testReport)** for PR 27359 at commit [`05c3f8d`](https://github.com/apache/spark/commit/05c3f8d56aff8fb01c7175c2abaab16cfa96a75b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22159/
Test PASSed.

SparkQA commented:
**[Test build #117399 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117399/testReport)** for PR 27359 at commit [`05c3f8d`](https://github.com/apache/spark/commit/05c3f8d56aff8fb01c7175c2abaab16cfa96a75b).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117399/
Test FAILed.

SparkQA commented:
**[Test build #117402 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117402/testReport)** for PR 27359 at commit [`3247cd4`](https://github.com/apache/spark/commit/3247cd418baca9cebd579726e11c1e0a379441d8).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22161/
Test PASSed.

SparkQA commented:
**[Test build #117402 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117402/testReport)** for PR 27359 at commit [`3247cd4`](https://github.com/apache/spark/commit/3247cd418baca9cebd579726e11c1e0a379441d8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117402/
Test PASSed.

zero323 commented:
Assuming that Windows test passes (it did previously), it is ready for review.

Unfortunately I couldn't figure out how to correctly install testthat 2.0 on Jenkins from inside test runner ‒ some configurations that worked locally (on official R-base docker image with 3.1.3, still failed here), and I don't want to throttle Jenkins anymore with my lousy attempts (sorry for that).

 So I decided to add fallback case, so it runs for now on both testthat 1.x and 2.x. It means that appveyor runs with 2.x and Jenkins with 1.x ‒ I guess it might be considered a good thing for now (@shaneknapp?). The only caveat is that we'll have to pay more attention to  AppVeyor builds, until Jenkins is upgraded, as it has become clear, that current setup can silence some errors.

CC @HyukjinKwon @felixcheung

dongjoon-hyun commented on R/pkg/tests/fulltests/test_sparkSQL.R:
```diff
@@ -848,24 +848,31 @@ test_that("collect() and take() on a DataFrame return the same number of rows an
   df <- read.df(jsonPath, "json")
   rdf <- collect(df)
   expect_true(is.data.frame(rdf))
-  expect_equal(rdf$name[1], markUtf8("안녕하세요"))
```
Ur, this is a required change for `testthat`?

zero323 replied:
This fails on Windows due to encoding issues, irrespective of testthat version. Similar to recursive call test case, failure is easy to reproduce (Windows, English locale,  CP1252 encoding) outside tests, but somehow is ignored in testthat 1.x.

@dongjoon-hyun The alternative to this is to write bytes explicitly, i.e.

```
test_that("collect() support Unicode characters", {

  lines <- markUtf8(c(
    '{"name": "안녕하세요"}',
    '{"name": "您好", "age": 30}',
    '{"name": "こんにちは", "age": 19}',
    '{"name": "Xin ch\xc3\xa0o"}'
  ))

  jsonPath <- tempfile(pattern = "sparkr-test", fileext = ".tmp")
  writeLines(lines, jsonPath, useBytes = TRUE)

  expected <- regmatches(lines, regexec('(?<="name": ").*?(?=")', lines, perl = TRUE))

  df <- read.df(jsonPath, "json")
  rdf <- collect(df)
  expect_true(is.data.frame(rdf))

  rdf$name <- markUtf8(rdf$name)
  expect_equal(rdf$name[1], expected[[1]])
  expect_equal(rdf$name[2], expected[[2]])
  expect_equal(rdf$name[3], expected[[3]])
  expect_equal(rdf$name[4], expected[[4]])

  df1 <- createDataFrame(rdf)
  expect_equal(
    collect(
      where(df1, df1$name == expected[[2]])
    )$name,
    expected[[2]]
  )
})
```

or skip test case on Windows, but as-is, this will and should fail on AppVeyor.

dongjoon-hyun replied:
Please make another PR for that. One PR should have one theme.
>  irrespective of testthat version

zero323 replied:
So you're saying that we should accept invalid test failures here? Or fix this beforehand on 1.x, even if problem is not visible with 1.x?

dongjoon-hyun replied:
Yes, right. That would be great.

dongjoon-hyun replied:
For Windows, @HyukjinKwon is the best person. So, he can override my comment.

zero323 replied:
Do you have any preference towards the method used?

In general I have no access to Windows setup on which I can test this across testthat (things vary, as 2.0 and latest 2.x versions come with different dependencies), R (we have 3.1 on Jenkins, latest on Windows builds) and encoding. So this is the best I came up with, in the limited time I had to tinker with the whole thing. But I am opened to other proposals.

dongjoon-hyun replied:
Ur, if you are unable to verify, please don't make a PR.
> I have no access to Windows setup

zero323 replied:
For the reference I've opened [SPARK-30645](https://issues.apache.org/jira/browse/SPARK-30645) to track this one.

dongjoon-hyun commented on R/pkg/tests/fulltests/test_sparkSQL.R:
```diff
@@ -848,24 +848,31 @@ test_that("collect() and take() on a DataFrame return the same number of rows an
+    Sys.getenv("SPARK_HOME"),
+    "R", "pkg", "tests", "fulltests", "data",
+    "test_utils_utf.json"
+  )
```
This new change looks irrelevant to `Update test that to >= 2.0.0`. I'd like to recommend you to revert this change.

zero323 replied:
Ditto.

dongjoon-hyun replied:
Ditto.

dongjoon-hyun commented on R/pkg/tests/fulltests/test_utils.R:
```diff
@@ -89,7 +89,10 @@ test_that("cleanClosure on R functions", {
     l$field[1, 1] <- 3  # Test for access operators `$`.
     res <- defUse + l$field[1, ]  # Test for def-use chain of "defUse", and "" symbol.
-    f(res)  # Test for recursive calls.
+    # Enable once SPARK-30629 is fixed
```
Does this fail with `testthat >= 2.0.0`?

zero323 replied:
@dongjoon-hyun It fails on any version, but is somehow silenced if we use testthat 1.x. I've already opened a JIRA ticket for ‒ [SPARK-30629](https://issues.apache.org/jira/browse/SPARK-30629). Can be easily reproduced outside tests.

dongjoon-hyun replied:
Let's fix SPARK-30629 first.

zero323 replied:
OK. Set SPARK-30629  as a blocker for SPARK-23435.

dongjoon-hyun replied:
Thanks, @zero323 . SPARK-30629 only happens at test that >= 2.0.0. It should be a blocker for this (SPARK-23435).

zero323 replied:
Dropping this commit from here.

zero323 replied:
CC @falaki

HyukjinKwon commented:
I am all good as long as we can use the latest `testthat`. This is a really annoying and rather critical problem because many R libraries nowsdays has `testthat` dependency.

SparkQA commented:
**[Test build #117406 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117406/testReport)** for PR 27359 at commit [`9cac55f`](https://github.com/apache/spark/commit/9cac55fb7c286939136a9bd06a9db3665959373f).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22165/
Test PASSed.

zero323 commented:
@HyukjinKwon 

> I am all good as long as we can use the latest `testthat`. This is a really annoying and rather critical problem because many R libraries nowsdays has `testthat` dependency.

Following @dongjoon-hyun advice I moved commits fixing test issues discovered here out this PR. Could you take a look at https://github.com/apache/spark/pull/27362 and see if it makes sense to you?

As of the other one, I'll try to debug this over next few days and see if I can find a fix, if not questions remains if it should be a blocker for testthat upgrade. It feel like a bit like sweeping things under the rug.

HyukjinKwon commented:
> Unfortunately I couldn't figure out how to correctly install testthat 2.0 on Jenkins from inside test runner ‒ some configurations that worked locally (on official R-base docker image with 3.1.3, still failed here), and I don't want to throttle Jenkins anymore with my lousy attempts (sorry for that).

@zero323, so basically you couldn't figure out CRAN check to work with R 3.1.3? If that's the case, does it work with the latest of R 3.4.x (at least in your local)?

Spark will deprecate R lower versions than 3.4 in Spark 3.0 and we will drop it right away after Spark 3.0 release. So, R version in Jenkins will increase to the latest of 3.4.x ideally.

HyukjinKwon commented on R/pkg/tests/run-all.R:
```diff
@@ -60,11 +59,23 @@ if (identical(Sys.getenv("NOT_CRAN"), "true")) {
-                         NULL,
-                         "summary")
+
+    # To be removed once testthat 1.x is removed from all builds
```
+1. I like the idea of keeping it to make the dev env change smoother. Also, we could decouple this PR and Jenkins R upgrade issue.

dongjoon-hyun replied:
+1. BTW, we had better file a new JIRA and make this comment as an IDed TODO like the following.
```
# TODO(SPARK-XXXXX) To be removed once testthat 1.x is removed from all builds
```

HyukjinKwon replied:
@zero323, can you handle this comment? Otherwise, It think it's good to go.

zero323 replied:
@HyukjinKwon Done 56b1ba966467f5677572d70fa5c4d75bc0c66789

zero323 commented:
> @zero323, so basically you couldn't figure out CRAN check to work with R 3.1.3? If that's the case, does it work with the latest of R 3.4.x (at least in your local or any Linux based OS)?

As long as fix for [SPARK-30645](https://issues.apache.org/jira/browse/SPARK-30645) is applied and test for cleanClosure on recursive ([SPARK-30629](https://issues.apache.org/jira/browse/SPARK-30629)) is disabled everything looks green, i.e.

- Passes all tests locally on R 3.6.2 / testthat 2.3.1
- It passes all tests on AppVeyor (I don't recall R version there, but testthat is 2.3.1).
- On Jenkins we fallback to 1.x so it is not really meaningful (I don't see how SPARK-30629 flies under the radar there), but all tests passed as well.

About _I couldn't figure_ ‒ I tried to `install.packages` inside `R/pkg/tests/run-all.R` in different combinations that seem to work locally, but it failed miserably on Jenkins (in general with R 3.1 used there we can target only `testthat` 2.0 or so, as later versions require 0.4, which in turn depends on R >= 3.2).

SparkQA commented:
**[Test build #117406 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117406/testReport)** for PR 27359 at commit [`9cac55f`](https://github.com/apache/spark/commit/9cac55fb7c286939136a9bd06a9db3665959373f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117406/
Test PASSed.

HyukjinKwon commented:
@zero323, do you mind if I ask to check R 3.4.x latest and `testthat` latest combination? After we drop R lower than 3.4, we will probably switch R version to 3.4.x latest in Jenkins. Then, I think this PR is almost good to go after https://issues.apache.org/jira/browse/SPARK-30629. I, @felixcheung or @shivaram might have to take another look though.

zero323 commented:
> @zero323, do you mind if I ask to check R 3.4.x latest and `testthat` latest combination? After we drop R lower than 3.4 (which will be after Spark 3.0 release since we already deprecated), we will probably switch R version to 3.4.x latest in Jenkins. Then, I think this PR is almost good to go after https://issues.apache.org/jira/browse/SPARK-30629. I, @felixcheung or @shivaram might have to take another look though.

Not at all, unless you ask for Windows build :) 

Seriously though, I had enough problems in category "works for me" with this thing, to be careful. Once https://github.com/apache/spark/pull/27363 I'll rebase, and make another local build in docker to be sure.

zero323 committed:
Switch to testthat >= 2.0.0

SparkQA commented:
**[Test build #117412 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117412/testReport)** for PR 27359 at commit [`05e8819`](https://github.com/apache/spark/commit/05e8819066b12961c6362ecfd70626470b6fe527).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22171/
Test PASSed.

dongjoon-hyun commented on appveyor.yml:
```diff
@@ -43,11 +43,8 @@ install:
-  # Therefore, we don't use devtools but installs it directly from the archive including its dependencies.
   - cmd: R -e "install.packages(c('crayon', 'praise', 'R6'), repos='https://cloud.r-project.org/')"
-  - cmd: R -e "install.packages('https://cloud.r-project.org/src/contrib/Archive/testthat/testthat_1.0.2.tar.gz', repos=NULL, type='source')"
+  - cmd: R -e "install.packages('testthat', repos='https://cloud.r-project.org/')"
```
Shall we add `testthat` to line 46 (https://github.com/apache/spark/pull/27359/files#diff-180360612c6b8c4ed830919bbb4dd459R46)?

HyukjinKwon replied:
+1 for @dongjoon-hyun's advice.

dongjoon-hyun commented on R/pkg/tests/run-all.R:
```diff
@@ -60,11 +59,23 @@ if (identical(Sys.getenv("NOT_CRAN"), "true")) {
+    }
+
+    test_runner("SparkR",
+           file.path(sparkRDir, "pkg", "tests", "fulltests"),
```
Indentation?

HyukjinKwon replied:
Should be:

```r
    test_runner("SparkR",
                file.path(sparkRDir, "pkg", "tests", "fulltests"),
                ...
```

to be consistent.

zero323 committed:
Apply style suggestions

SparkQA commented:
**[Test build #117412 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117412/testReport)** for PR 27359 at commit [`05e8819`](https://github.com/apache/spark/commit/05e8819066b12961c6362ecfd70626470b6fe527).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117412/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22174/
Test PASSed.

dongjoon-hyun commented:
@zero323 . Please update the PR description. For example, the followings?
```
Disables unit tests for faulty code (SPARK-30629)
Modify Unicode handling test case to be handled properly independent of OS.
```

SparkQA commented:
**[Test build #117415 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117415/testReport)** for PR 27359 at commit [`69544ae`](https://github.com/apache/spark/commit/69544ae1a0bd72112f65354dc772a61dbcff0fdf).

SparkQA commented:
**[Test build #117415 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117415/testReport)** for PR 27359 at commit [`69544ae`](https://github.com/apache/spark/commit/69544ae1a0bd72112f65354dc772a61dbcff0fdf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117415/
Test PASSed.

zero323 committed:
Replace installed.packages with packageVersion

SparkQA commented:
**[Test build #117426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117426/testReport)** for PR 27359 at commit [`c7ed64a`](https://github.com/apache/spark/commit/c7ed64af9e697b3619779857dd820832176b3be3).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22185/
Test PASSed.

SparkQA commented:
**[Test build #117426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117426/testReport)** for PR 27359 at commit [`c7ed64a`](https://github.com/apache/spark/commit/c7ed64af9e697b3619779857dd820832176b3be3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117426/
Test PASSed.

zero323 commented:
@HyukjinKwon 

> @zero323, do you mind if I ask to check R 3.4.x latest and testthat latest combination

I'd say we're good:

[![asciicast](https://asciinema.org/a/xiIOy6OcntE6hxNXQwI7vcUl0.svg)](https://asciinema.org/a/xiIOy6OcntE6hxNXQwI7vcUl0)

Additionally to local builds this gives us:

- R 3.1.x, `testthat` 1.0.2 on Linux (Jenkins)
- R 3.4.3, `testthat` 2.3.1 on Linux (docker build)
- R 3.6.2, `testthtat` 2.3.1 on Windows

zero323 commented:
> Please update the PR description. For example, the followings?

All done @dongjoon-hyun

dongjoon-hyun commented:
@zero323 . Thank you for the screencast. However, it skipped all arrow related tests.

zero323 commented:
> @zero323 . Thank you for the screencast. However, it skipped all arrow related tests. Please playback the screencast.

Unfortunately R arrow is not standalone package (like Python one) and it requires system packages with C++ bindings (installing `arrow` package is not sufficient), And that's dependency hell as these R images (there still more stable than R-base ones) are not really designed for updates. So skipped Arrow tests are expected.

I don't think that really affects the results though, as primary concern was CRAN tests and overall process, and Arrow related code hasn't been touched.

dongjoon-hyun commented:
We cannot say `We're good` when we know something wrong. I'd like to recommend you to mention what you've done clearly. That's enough. Please note that I'm supporting your effort on this PR. Otherwise, I'll not chim in here to add comments.
> So skipped Arrow tests are expected.

zero323 commented:
> Please note that I'm supporting your effort on this PR. Otherwise, I'll not chim in here to add comments.

Thank you, I appreciate that. 

In general, full reproducible is defined by the Dockerfile which is shown at the begging,  but to put it here for reference 

```
FROM rocker/verse:3.4.3
RUN apt-get update \
    && apt-get install -y --no-install-recommends gpg openjdk-8-jdk-headless \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*ce

RUN wget -qO- https://keybase.io/zero323/pgp_keys.asc | gpg --import
RUN git clone --depth 1 --branch SPARK-23435 https://github.com/zero323/spark.git
WORKDIR spark
RUN git rev-parse HEAD
RUN git verify-commit -v HEAD
RUN build/mvn -DskipTests -Phive -Psparkr clean package
RUN R --version
RUN R -e "install.packages(c('e1071', 'praise'))"
RUN R -e "install.packages('testthat', repos='https://cloud.r-project.org/'); packageVersion('testthat'); sessionInfo()"
RUN R/create-rd.sh
RUN R/create-docs.sh
RUN R/check-cran.sh
RUN R/run-tests.sh
```
It can be re-run to confirm that it reflects current state of things.

As show in the cast, build are done directly from this head of this branch (signature is verified) and no changes to the codebase, beyond what is proposed in this PR (and we don't touch any Arrow related components here at all).

As of skipping Arrow tests - that's default behavior defined in respective test for example here

https://github.com/apache/spark/blob/43d9c7e7e57749ee611e0c97781a71a0645b5e9b/R/pkg/tests/fulltests/test_sparkSQL_arrow.R#L25

and following lines. So it is neither failure or result of any source modification. 

Can we make arrow tests run? Possibly, but:

- R Arrow package is not present in snapshot repositories used by rocker images. Installing testthat from https://cloud.r-project.org, already pushed things a lot. Additionally some transitive dependencies have hidden version bounds.
- C++ Arrow bindings would require external system repositories, which can break decencies for R.
- Using other images (let's say official R-base) is not an option, as we need Tex as well as OpenSSL and Curl dev libraries and this will either break or require update of R beyond 2.4 (at least it did for other build configurations I considered).

At this point Spark has no coverage for any intermediate R version (Jenkins runs 3.1 and then we have almost eight years of releases worth gap to 3.6 on AppVeyor), not to mention version-OS combinations. That's troubling, and as work related to this PR shown, can miss obvious errors.  but  not something that can be really addressed by running ad-hoc tests outside project infrastructure.

Anyway... If you have specific concerns about the process used here, and you suspect that proposed changes can lead to problems in the future, I'll do my best to address these.

dongjoon-hyun commented:
? @zero323 . It seems that you missed my point. I advised like the following.
>  I'd like to recommend you to mention what you've done clearly. That's enough.

Let me rephrase my words. "In the PR description, write that you didn't run the full test. Especially Arrow tests are skipped". It was my request.

HyukjinKwon reviewed (approved):
LGTM.

Arrow is a valid point although AppVeyor (with R latest and CRAN disabled) tests it. It should be best just to test it out if that's feasible considering the importance of this PR and that changing env in Jenkins is difficult to revert if there's anything wrong.

If there's some technical difficulties, I am okay without testing out because the required codes change `expect_true(TRUE)` in the tests seem not required for Arrow tests logically at least.

HyukjinKwon reviewed (commented):
Okay, given https://github.com/apache/spark/pull/27359#issuecomment-578581851, I am okay with not trying Arrow installation out. 

@shaneknapp made it installing Arrow on R 3.1 (surprisingly ..). Although it's a bit of a while ago and Arrow R was at the early stage, I remember I badly failed to install Arrow in some low R versions.

I think we can verify it later when we actually upgrade R version. Seems it should be easy to fix if there'a anything wrong with Arrow tests.

Also, I found a test case that has a similar structure with Arrow test case as a reference:
https://github.com/apache/spark/blob/d69ed9afdf2bd8d03aaf835292b92692ec8189e9/R/pkg/tests/fulltests/test_sparkSQL.R#L3327-L3376

So I suspect it's good enough to go (except one minor comment https://github.com/apache/spark/pull/27359#discussion_r371046762)

zero323 committed:
Add JIRA ticket to TODO removal comment

Signed-off-by: zero323 <mszymkiewicz@gmail.com>

SparkQA commented:
**[Test build #117484 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117484/testReport)** for PR 27359 at commit [`56b1ba9`](https://github.com/apache/spark/commit/56b1ba966467f5677572d70fa5c4d75bc0c66789).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/22242/
Test PASSed.

zero323 commented:
@dongjoon-hyun 

> ? @zero323 . It seems that you missed my point. I advised like the following.
> 
> > I'd like to recommend you to mention what you've done clearly. That's enough.
> 
> Let me rephrase my words. "In the PR description, write that you didn't run the full test. Especially Arrow tests are skipped". It was my request.

Seems like. I guess there is something lost in translation here

> We cannot say We're good when we know something wrong

Especially what we consider "wrong" in this context, hence my elaborate response. Which I think is still valid ‒ overall test matrix is tiny compared to what is officially supported. Sadly Arrow makes things worse as

- It is not Maven dependency.
- Building from source is pain, and repositories don't provide archival versions.
- Depending on OS, there are also additional dependencies.
- Breaking API changes are not unusual (for example it changed between first 3.0.0 preview and the second), leading to rather [cryptic errors](https://twitter.com/_zero323/status/1202005430471450625).

It would be great if we could either isolate that (I played a bit with Anaconda for that, and results are quite promising, but I am not there yet, not to mention it won't help with all the tech. debt caused by supporting long outdated R and Python versions), or find a way to bundle within Maven builds.

For now I published [zero323/sparkr-build-sandbox](https://github.com/zero323/sparkr-build-sandbox) ([docker.com/r/zero323/sparkr-build-sandbox](https://hub.docker.com/r/zero323/sparkr-build-sandbox)) - not exactly what is needed, but does the trick for now.

Please check updated description and let me know if that addresses your concerns.

SparkQA commented:
**[Test build #117484 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117484/testReport)** for PR 27359 at commit [`56b1ba9`](https://github.com/apache/spark/commit/56b1ba966467f5677572d70fa5c4d75bc0c66789).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117484/
Test PASSed.

dongjoon-hyun commented:
Sorry, but it doesn't address my concerns.

> Please check updated description and let me know if that addresses your concerns.

<img width="565" alt="Screen Shot 2020-01-28 at 2 14 29 PM" src="https://user-images.githubusercontent.com/9700541/73310146-82b71780-41d8-11ea-8db7-2662bfb544a7.png">

I ask you the following. In the above, I cannot see the following content. Even, you completely hide the word `arrow`.
> "In the PR description, write that you didn't run the full test. Especially Arrow tests are skipped". It was my request.

dongjoon-hyun reviewed (changes requested):
@HyukjinKwon . Sorry, but you know that I'm not against that @zero323 skipped some tests due to the difficulty. I fully understand that it's sometime unfair to ask that to the contributors. However, the author of PR should mention what he've done in the PR description correctly.

I'm not sure why @zero323 adds many comments about this difficulty, but he didn't mention any of those difficulty into the PR description which will become a permanent commit log. Could you advise him about that? I already add the same comments repeatedly. So, it's too much for me.

Or, @HyukjinKwon . You can merge this PR as is. It's up to you. I'll not be against your decision.

zero323 commented:
@dongjoon-hyun 

> "In the PR description, write that you didn't run the full test. 

At this point all tests *did* run across multiple additional R versions. This is both documented in a persistent and verifiable way, and reproducible.

zero323 commented:
Retest this please.

HyukjinKwon commented:
@zero323, mind adding a couple of words to describe `arrow` was not installed due to technical difficulties, into the PR description? Seems like there was some miscommunication here.

It's fine to describe what happened during manual test. No one can't be flawless during such tests; it's better to describe it thoroughly for people who follow up later.

SparkQA commented:
**[Test build #117498 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117498/testReport)** for PR 27359 at commit [`56b1ba9`](https://github.com/apache/spark/commit/56b1ba966467f5677572d70fa5c4d75bc0c66789).

zero323 commented:
@HyukjinKwon 

> mind adding a couple of words to describe arrow was not installed due to technical difficulties, into the PR description?

I was actually able to put all the dependencies together, including Arrow. 

> Seems like there was some miscommunication here.

Indeed. I guess I should have skipped explanations.

Anyway, I am happy to close this, and leave for someone more articulate to continue. Everything should be in good shape here from technical perspective, and linked images should provide good starting point for anyone willing to continue. Let's just see if AppVeyor failure is a false positive. Seems like [the last run](https://ci.appveyor.com/project/ApacheSoftwareFoundation/spark/builds/30408028?fullLog=true) failed without clear indication why.

Thanks for your work @HyukjinKwon @dongjoon-hyun and sorry for any misunderstandings on my side.

HyukjinKwon reviewed (approved):
Alright, then I am going to merge this. LGTM.

SparkQA commented:
**[Test build #117498 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/117498/testReport)** for PR 27359 at commit [`56b1ba9`](https://github.com/apache/spark/commit/56b1ba966467f5677572d70fa5c4d75bc0c66789).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/117498/
Test PASSed.

HyukjinKwon commented:
Merged to master.

HyukjinKwon commented:
@shaneknapp TR;DL: Now SparkR supports both testthat 1.0.2 and 2.0.0. I will backport this to branch-2.4 at #27379 as well.

So, as a reminder:
- We should upgrade testthat in Jenkins to 2.0.0 at any earliest time you're available
- We should upgrade R version to 3.4.x after Spark 3.0 release.

shaneknapp commented:
@HyukjinKwon will do.  the upgrade won't happen until next week.

shaneknapp commented:
> * We should upgrade testthat in Jenkins to 2.0.0 at any earliest time you're available

should be easy.

> * We should upgrade R version to 3.4.x after Spark 3.0 release.

won't be easy.  will be disruptive.
